### PR TITLE
FoliagePlacer mapping and doc, Tree mapping, and a lot of misc feature mapping

### PR DESCRIPTION
**Pull request by jaskarth**
Originally this was just supposed to be for trees and foliage placers but a got a little bit carried away lol

### CODE REVIEW

File: mappings/net/minecraft/util/math/BlockPos.mapping [hunk 1]
```diff
@@ -98,6 +98,7 @@ CLASS net/minecraft/class_2338 net/minecraft/util/math/BlockPos
 		ARG 1 pos2
 	METHOD method_23627 stream (Lnet/minecraft/class_3341;)Ljava/util/stream/Stream;
 		ARG 0 box
+	METHOD method_25503 toMutable ()Lnet/minecraft/class_2338$class_2339;
```
Comment (liach) [inline mappings/net/minecraft/util/math/BlockPos.mapping]: Not really, this is `mutableCopy` as the `Mutable` subclass does not have an implementation that returns itself.

File: mappings/net/minecraft/util/math/BlockPos.mapping [hunk 2]
```diff
@@ -139,3 +140,10 @@ CLASS net/minecraft/class_2338 net/minecraft/util/math/BlockPos
 			ARG 2 x
 			ARG 3 y
 			ARG 4 z
+		METHOD method_25504 setOffset (Lnet/minecraft/class_2382;III)Lnet/minecraft/class_2338$class_2339;
+			ARG 1 pos
+			ARG 2 x
+			ARG 3 y
+			ARG 4 z
+		METHOD method_25505 setOffset (Lnet/minecraft/class_2382;Lnet/minecraft/class_2350;)Lnet/minecraft/class_2338$class_2339;
```
Comment (liach) [inline mappings/net/minecraft/util/math/BlockPos.mapping]: I propose renaming such `setOffset` to `move`

File: mappings/net/minecraft/world/gen/feature/TreeFeatureConfig.mapping [hunk 3]
```diff
@@ -3,13 +3,15 @@ CLASS net/minecraft/class_4643 net/minecraft/world/gen/feature/TreeFeatureConfig
...
 	FIELD field_21291 baseHeight I
+	FIELD field_21593 skipFluidCheck Z
 	METHOD <init> (Lnet/minecraft/class_4651;Lnet/minecraft/class_4651;Ljava/util/List;I)V
 		ARG 1 trunkProvider
 		ARG 2 leavesProvider
 		ARG 3 decorators
 		ARG 4 baseHeight
 	METHOD method_23444 deserialize (Lcom/mojang/datafixers/Dynamic;)Lnet/minecraft/class_4643;
 		ARG 0 configDeserializer
+	METHOD method_23916 setSkipFluidCheck ()V
```
Comment (liach) [inline mappings/net/minecraft/world/gen/feature/TreeFeatureConfig.mapping]:
```suggestion
	METHOD method_23916 ignoreFluidCheck ()V
```

File: mappings/net/minecraft/world/gen/foliage/FoliagePlacer.mapping [hunk 4]
```diff
@@ -6,11 +6,55 @@ CLASS net/minecraft/class_4647 net/minecraft/world/gen/foliage/FoliagePlacer
...
+		ARG 1 trunkHeight
+		ARG 2 baseHeight
+			COMMENT This method is used to ensure that a tree can place foliage when it generates.
+			COMMENT
+			COMMENT It runs for every y-level of the tree being generated.
+			COMMENT
+			COMMENT @param trunkHeight the height of the trunk
+			COMMENT @param baseHeight the height of the hull tree
+			COMMENT @param radius the radius of the FoliagePlacer
+			COMMENT @param currentTreeHeight the current y-level of the tree being tested.
```
Comment (liach) [inline mappings/net/minecraft/world/gen/foliage/FoliagePlacer.mapping]: Why are you commenting on param 2 than the method itself? also `@param` tags should be replaced by comments on individual parameter themselves.

File: mappings/net/minecraft/world/gen/foliage/FoliagePlacer.mapping [hunk 5]
```diff
@@ -6,11 +6,55 @@ CLASS net/minecraft/class_4647 net/minecraft/world/gen/foliage/FoliagePlacer
...
+		ARG 3 radius
+		ARG 4 currentTreeHeight
 	METHOD method_23448 generate (Lnet/minecraft/class_3747;Ljava/util/Random;Lnet/minecraft/class_4640;IIILnet/minecraft/class_2338;Ljava/util/Set;)V
+		COMMENT This is the main method used to generate foliage.
+		COMMENT
+		COMMENT @param baseHeight the height of the full tree
+		COMMENT @param trunkHeight the height of just the trunk, or the part of the tree without leaves
+		COMMENT @param radius the radius of the foliage
```
Comment (liach) [inline mappings/net/minecraft/world/gen/foliage/FoliagePlacer.mapping]: Move `@param` to param themselves

File: mappings/net/minecraft/world/gen/foliage/FoliagePlacer.mapping [hunk 6]
```diff
@@ -6,11 +6,55 @@ CLASS net/minecraft/class_4647 net/minecraft/world/gen/foliage/FoliagePlacer
...
+		ARG 7 radius
+		ARG 8 leaves
+	METHOD method_23450 placeLeaves (Lnet/minecraft/class_3747;Ljava/util/Random;Lnet/minecraft/class_2338;Lnet/minecraft/class_4640;Ljava/util/Set;)V
+		ARG 1 world
+		ARG 2 random
+		ARG 3 pos
+		ARG 4 config
+		ARG 5 leaves
+			COMMENT
+	METHOD method_23451 invalidForLeaves (Ljava/util/Random;IIIII)Z
```
Comment (liach) [inline mappings/net/minecraft/world/gen/foliage/FoliagePlacer.mapping]:
```suggestion
	METHOD method_23451 isInvalidForLeaves (Ljava/util/Random;IIIII)Z
```

File: mappings/net/minecraft/world/gen/feature/VoidStartPlatformFeature.mapping [hunk 7]
```diff
@@ -1 +1,8 @@
 CLASS net/minecraft/class_3217 net/minecraft/world/gen/feature/VoidStartPlatformFeature
+	FIELD field_19241 START_BLOCK Lnet/minecraft/class_2338;
+	FIELD field_19242 START_CHUNK Lnet/minecraft/class_1923;
+	METHOD method_20403 distance (IIII)I
```
Comment (liach) [inline mappings/net/minecraft/world/gen/feature/VoidStartPlatformFeature.mapping]:
```suggestion
	METHOD method_20403 getDistance (IIII)I
```

**Review by liach — COMMENTED** [review]
Looks good overall. A few minor problems.

File: mappings/net/minecraft/world/gen/foliage/FoliagePlacer.mapping [hunk 8]
```diff
@@ -7,16 +7,16 @@ CLASS net/minecraft/class_4647 net/minecraft/world/gen/foliage/FoliagePlacer
...
 		ARG 3 type
 	METHOD method_23447 getRadiusForPlacement (IIII)I
+		COMMENT This method is used to ensure that a tree can place foliage when it generates.
+		COMMENT
+		COMMENT It runs for every y-level of the tree being generated.
+		COMMENT
+		COMMENT @param trunkHeight the height of the trunk
+		COMMENT @param baseHeight the height of the hull tree
+		COMMENT @param radius the radius of the FoliagePlacer
+		COMMENT @param currentTreeHeight the current y-level of the tree being tested.
```
Comment (liach) [inline mappings/net/minecraft/world/gen/foliage/FoliagePlacer.mapping]: Yeah please place the `@param` tags' contents to individual param themselves.

File: mappings/net/minecraft/world/gen/foliage/FoliagePlacer.mapping (hunk 4 above)
Comment (jaskarth) [reply mappings/net/minecraft/world/gen/foliage/FoliagePlacer.mapping]: oops

**Review by liach — APPROVED** [review]

**Review by Runemoro — APPROVED** [review]

File: mappings/net/minecraft/world/gen/feature/MegaTreeFeature.mapping [hunk 9]
```diff
@@ -4,12 +4,32 @@ CLASS net/minecraft/class_3096 net/minecraft/world/gen/feature/MegaTreeFeature
...
+		ARG 2 config
 	METHOD method_13525 replaceGround (Lnet/minecraft/class_3747;Lnet/minecraft/class_2338;)Z
 		ARG 1 world
 		ARG 2 pos
 	METHOD method_13526 makeRoundLeafLayer (Lnet/minecraft/class_3747;Ljava/util/Random;Lnet/minecraft/class_2338;ILjava/util/Set;Lnet/minecraft/class_3341;Lnet/minecraft/class_4643;)V
+		ARG 1 world
+		ARG 3 pos
+		ARG 4 radius
+		ARG 5 leaves
+		ARG 6 ox
```
Comment (Juuxel) [inline mappings/net/minecraft/world/gen/feature/MegaTreeFeature.mapping]:
```suggestion
		ARG 6 box
```

File: mappings/net/minecraft/world/gen/foliage/FoliagePlacer.mapping [hunk 10]
```diff
@@ -6,11 +6,52 @@ CLASS net/minecraft/class_4647 net/minecraft/world/gen/foliage/FoliagePlacer
 		ARG 1 radius
 		ARG 2 randomRadius
 		ARG 3 type
+	METHOD method_23447 getRadiusForPlacement (IIII)I
+		COMMENT This method is used to ensure that a tree can place foliage when it generates.
+		COMMENT
+		COMMENT It runs for every y-level of the tree being generated.
```
Comment (Juuxel) [inline mappings/net/minecraft/world/gen/foliage/FoliagePlacer.mapping]:
```suggestion
		COMMENT <p>It runs for every y-level of the tree being generated.
```

**Review by Juuxel — CHANGES_REQUESTED** [review]

**Review by Juuxel — APPROVED** [review]

**Pull request merged**